### PR TITLE
Add Keras/TorchNN tutorial

### PR DESCRIPTION
**Pull request by trbromley**
This tutorial shows how PennyLane QNodes can be integrated with Keras and torch.nn using the newly added `KerasLayer` and `TorchLayer` functionality in the `qnn` module.

### CODE REVIEW

**Comment by josh146** [comment]
<h3>Website build</h3>
<strong>Commit:</strong> 709c8bbfa3eb7d49abdb8cb4ce0d3111068e70b9

<strong>Circle build number:</strong> 1169

<strong>Built <code>demos</code> directory:</strong> <a href=https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_demos.zip>https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_demos.zip</a>

<strong>Backreferences zip:</strong> <a href=https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_backreferences.zip>https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_backreferences.zip</a>

<strong>Website zip:</strong> <a href=https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_html.zip>https://1169-214003948-gh.circle-artifacts.com/0/tmp/qml_html.zip</a>

<strong>Website url:</strong> <a href=https://1169-214003948-gh.circle-artifacts.com/0/_build/html/index.html>https://1169-214003948-gh.circle-artifacts.com/0/_build/html/index.html</a>

<em>Please double check the rendered website build to make sure everything is correct.</em>

File: demonstrations/tutorial_qnn_module.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
+r"""
+Hybrid quantum-classical ML using Keras and PyTorch
```
Comment (mariaschuld) [inline demonstrations/tutorial_qnn_module.py:2]: I wonder if the title could be made more descriptive and do the great work behind this tutorial justice. What about: "Adding quantum nodes as Keras or Pytorch Layers"? I'm sure people will understand what "layers" means here...

File: demonstrations/tutorial_qnn_module.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    :property="og:description": Learn how to create hybrid ML models in PennyLane using the Keras and PyTorch interfaces
+    :property="og:image": https://pennylane.ai/qml/_images/qnn21.png
+
+In PennyLane, variational quantum circuits are treated as :doc:`quantum nodes
+<../glossary/hybrid_computation>` (QNodes) that can be thought of simply as functions with a
+defined output and gradient for a given input. QNodes can be combined with classical nodes to
+form a hybrid model, which can be trained using the TensorFlow, PyTorch or Autograd/NumPy
+:doc:`interfaces <introduction/interfaces>`.
+
+Composing computational nodes is a fundamental tool of any machine learning package. However,
```
Comment (mariaschuld) [inline demonstrations/tutorial_qnn_module.py:15]: Maybe there is a more spicy & focused beginning? Along the lines of: 

"In Pytorch, you add a layer to a NN like this (code example). In TensorFlow, Keras provides the same functionality like (this). PennyLane allows you to use a quantum algorithm or "quantum node" as this building block..."

Then everyone immediately knows what you are trying to do, especially if people come from classical ML.

By the way, I would also prefer duplicating this tut, once with Keras and once with TF.

File: demonstrations/tutorial_qnn_module.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+#
+# With the QNode defined, we are ready to interface with `Keras <https://keras.io/>`__ or
+# `torch.nn <https://pytorch.org/docs/stable/nn.html>`__. This is achieved using the
+# :class:`~pennylane.qnn.KerasLayer` and :class:`~pennylane.qnn.TorchLayer` classes of the
+# :mod:`~pennylane.qnn` module, which convert the QNode to the elementary building block of these
+# high-level frameworks: a *layer*. We shall see in the following how the resultant layer can be
+# combined with other well-known neural network layers to form a hybrid model.
+#
+# We must first define the ``weight_shapes`` dictionary. Recall that all of
+# the arguments of the QNode except for the ``inputs``-named argument are treated as trainable
```
Comment (mariaschuld) [inline demonstrations/tutorial_qnn_module.py:86]:
```suggestion
# the arguments of the QNode (except the ones named ``inputs``) are treated as trainable
```

File: demonstrations/tutorial_qnn_module.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+# shape of each trainable weight for them to be initialized. The ``weight_shapes`` dictionary
+# maps from the argument names of the QNode to corresponding shapes:
+
+n_layers = 3
+weight_shapes = {"weights": (n_layers, n_qubits, 3)}
+
+###############################################################################
+# In our example, the ``weights`` argument of the QNode is trainable and has shape given by
+# ``(n_layers, n_qubits, 3)`` which are passed to
+# :func:`~pennylane.templates.layers.StronglyEntanglingLayers`.
```
Comment (mariaschuld) [inline demonstrations/tutorial_qnn_module.py:98]: Can we use ``BasicEntanglingLayers`` here? I am still trying to steer tutorials away from ``StronglyEntangling``, because the architecture is so much more complex and also just a heuristic (i.e. no one knows if they really entangle stronger than others).

File: demonstrations/tutorial_qnn_module.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+@qml.qnode(dev)
+def qnode(inputs, weights):
+    qml.templates.AngleEmbedding(inputs, wires=range(n_qubits))
+    qml.templates.StronglyEntanglingLayers(weights, wires=range(n_qubits))
+    return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]
+
+###############################################################################
+# Note that we do not need to specify the ``interface`` argument in the :func:`~pennylane.qnode`
+# decorator.
```
[the remark this answers is not in the document]
Comment (josh146) [reply demonstrations/tutorial_qnn_module.py:73]: @trbromley and I have discussed this online, and there is a bugfix that needs to be made to PL core to fix this functionality.

**Comment by josh146** [comment]
> I agree with Josh that it would read a lot easier as two tutorials, one with TF and one with Keras.

Another argument for this is that most users already come from a specific framework, and want to know how PL integrates with either keras _or_ torch.nn, but not both.

**Comment by mariaschuld** [comment]
Hey @trbromley , let us know when you are ready for re-review!

**Comment by trbromley** [comment]
Thanks @mariaschuld! I started making some changes but still not done yet, this one got put as a bit low priority.

**Comment by co9olguy** [comment]
@trbromley just checking in on this one. Any updates planned?

**Comment by trbromley** [comment]
@co9olguy, I had a brief look last Friday but decided to hold off to see if anything changes from the `quantum-wires` refactor, otherwise the tutorials might end up out of date pretty quickly.

**Comment by co9olguy** [comment]
Isn't the quantum wires refactor already finished?

**Comment by trbromley** [comment]
Sorry, I meant `quantum-tape` refactor (edited comment)

**Comment by trbromley** [comment]
[ch1802]

**Review by trbromley — COMMENTED** [review]
Thanks @josh146 @mariaschuld for the great comments! I've tried to follow them as close as possible and think the tutorial(s) are better now. The big changes are some improvements to the introduction and splitting out into two tutorials.

File: demonstrations/tutorial_qnn_module_tf.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+###############################################################################
+# The model can then be trained using `model.fit()
+# <https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit>`__.
+#
+# **What if we want to add a quantum layer to our model?** This is possible in PennyLane:
+# :doc:`QNodes <../glossary/hybrid_computation>` can be converted into Keras layers and combined
+# with the wide range of built-in classical
+# `layers <https://www.tensorflow.org/api_docs/python/tf/keras/layers>`__ to create truly hybrid
+# models. This tutorial will guide you through a simple example to show you how it's done!
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:30]: This is great; makes it easily searchable if a user has that question, and states directly what the goal of the demo is

File: demonstrations/tutorial_qnn_module_tf.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+#
+# .. note::
+#
+#     A similar demo explaining how to
+#     :doc:`turn quantum nodes into Torch layers <tutorial_qnn_module_torch>`
+#     is also available.
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:36]: :+1:

File: demonstrations/tutorial_qnn_module_tf.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+# :doc:`default.qubit <code/api/pennylane.devices.default_qubit.DefaultQubit>` simulator and
+# operations from the :doc:`templates <introduction/templates>` module.
+
+import pennylane as qml
+
+n_qubits = 2
+dev = qml.device("default.qubit", wires=n_qubits)
+
+
+@qml.qnode(dev)
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:76]:
```suggestion
n_qubits = 2
dev = qml.device("default.qubit", wires=n_qubits)

@qml.qnode(dev)
```

File: demonstrations/tutorial_qnn_module_tf.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+#
+# Creating a hybrid model
+# -----------------------
+#
+# Let's create a basic three-layered hybrid model consisting of:
+#
+# 1. A 2-neuron fully connected classical layer
+# 2. Our 2-qubit QNode converted into a layer
+# 3. Another 2-neuron fully connected classical layer
+# 4. A softmax activation to convert to a probability vector
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:123]:
```suggestion
# 1. a 2-neuron fully connected classical layer,
# 2. our 2-qubit QNode converted into a layer,
# 3. another 2-neuron fully connected classical layer, and
# 4. a softmax activation to convert to a probability vector.
```

Comment (josh146) [reply demonstrations/tutorial_qnn_module_tf.py:123]: (minor and stylistic, feel free not to accept)

File: demonstrations/tutorial_qnn_module_tf.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+#
+# A diagram of the model can be seen in the figure below.
+#
+# .. figure:: /demonstrations/qnn_module/qnn_keras.png
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:127]: The diagram looks much better now!

File: demonstrations/tutorial_qnn_module_tf.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+#
+# We can construct the model using the
+# `Sequential <https://www.tensorflow.org/api_docs/python/tf/keras/Sequential>`__ API:
+
+clayer_1 = tf.keras.layers.Dense(2)
+clayer_2 = tf.keras.layers.Dense(2, activation="softmax")
+model = tf.keras.models.Sequential([clayer_1, qlayer, clayer_2])
+
+###############################################################################
+# Constructing hybrid models is easy!
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:139]: I remember reading somewhere that you should never say something is easy/straightforward during a tutorial, as a reader who is finding it difficult to follow will read that sentence and consciously disagree 😆 Although I am guilty of also calling things easy in my demos.

File: demonstrations/tutorial_qnn_module_tf.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+###############################################################################
+# As a final step, let's train the model to check if it's working:
+
+opt = tf.keras.optimizers.SGD(learning_rate=0.2)
+model.compile(opt, loss="mae", metrics=["accuracy"])
+
+fitting = model.fit(X, y_hot, epochs=6, batch_size=5, validation_split=0.25, verbose=2)
+
+###############################################################################
+# Great! We've mastered the basics of constructing hybrid classical-quantum models using
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_tf.py:216]: I was kind of expecting to see a plot at the end, showing how well the classification did!

File: demonstrations/tutorial_qnn_module_torch.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+# ------------------------------
+#
+# Let us begin by choosing a simple dataset and problem to allow us to focus on how the hybrid
+# model is constructed. Our objective is to classify points generated from scikit-learn's
+# binary-class
+# `make_moons() <https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_moons.html>`__ dataset:
+
+import matplotlib.pyplot as plt
+from sklearn.datasets import make_moons
+import tensorflow as tf
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_torch.py:47]:
```suggestion
```

File: demonstrations/tutorial_qnn_module_torch.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+# two-qubit QNode using the
+# :doc:`default.qubit <code/api/pennylane.devices.default_qubit.DefaultQubit>` simulator and
+# operations from the :doc:`templates <introduction/templates>` module.
+
+import pennylane as qml
+
+n_qubits = 2
+dev = qml.device("default.qubit", wires=n_qubits)
+
+
```
Comment (josh146) [inline demonstrations/tutorial_qnn_module_torch.py:75]:
```suggestion
n_qubits = 2
dev = qml.device("default.qubit", wires=n_qubits)

```

File: demos_basics.rst [hunk 15]
```diff
@@ -132,6 +132,18 @@ hardware.
     :description: :doc:`demos/tutorial_vqe_uccsd_obs`
     :tags: chemistry
 
+.. customgalleryitem::
+    :tooltip: Learn how to create hybrid ML models using Keras
+    :figure: _static/Keras_logo.png
```
Comment (josh146) [inline demos_basics.rst:137]: I like the new thumbnails 👍

**Review by josh146 — APPROVED** [review]
Reads really well now Tom! I reviewed the Keras one first, most of my comments there also apply to the Torch one. But most were minor!

File: demonstrations/tutorial_qnn_module_torch.py (hunk 13 above)
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_torch.py:47]: Ah this was used for the `y_hot`. I found the numpy-based implementations a bit unclear: https://stackoverflow.com/questions/29831489/convert-array-of-indices-to-1-hot-encoded-numpy-array

File: demonstrations/tutorial_qnn_module_torch.py (hunk 14 above)
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_torch.py]: Fixed

File: demonstrations/tutorial_qnn_module_tf.py (hunk 12 above)
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_tf.py:212]: The best thing would be to plot either the decision boundary or just the predictions on the dataset. However, the decision boundary is linear. I suspect this can be fixed by making the model more interesting, but I'm worried that showing the plot might end up being more of a distraction if we have to explain why it's linear.

**Review by trbromley — COMMENTED** [review]
Thanks @josh146!

File: demonstrations/tutorial_qnn_module_torch.py (hunk 13 above)
Comment (josh146) [reply demonstrations/tutorial_qnn_module_torch.py:47]: Oh 🤔 Is there a torch native way of doing it? Feels a bit strange for the demo to depend on TF _and_ Torch 🤔

File: demonstrations/tutorial_qnn_module_tf.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+# -----------------------
+#
+# Let's create a basic three-layered hybrid model consisting of:
+#
+# 1. a 2-neuron fully connected classical layer
+# 2. our 2-qubit QNode converted into a layer
+# 3. another 2-neuron fully connected classical layer
+# 4. a softmax activation to convert to a probability vector
+#
+# A diagram of the model can be seen in the figure below.
```
Comment (glassnotes) [inline demonstrations/tutorial_qnn_module_tf.py:123]: I like how this part is structured, the image, code, and text describing the network all viewable at the same time. :100:

File: demonstrations/tutorial_qnn_module_torch.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
+"""
+Turning quantum nodes into Torch Layers
+=======================================
+
+.. meta::
+    :property="og:description": Learn how to create hybrid ML models in PennyLane using Torch
+    :property="og:image": https://pennylane.ai/qml/_images/PyTorch_icon.png
+
```
Comment (glassnotes) [inline demonstrations/tutorial_qnn_module_torch.py:8]:
```suggestion

.. related::

   tutorial_qnn_module_tf Turning quantum nodes into Keras Layers
```

File: demonstrations/tutorial_qnn_module_torch.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    avg_loss = running_loss / batches
+    print("Average loss over epoch {}: {:.4f}".format(epoch + 1, avg_loss))
+
+y_pred = model(X)
+predictions = torch.argmax(y_pred, axis=1).detach().numpy()
+
+correct = [1 if p == p_true else 0 for p, p_true in zip(predictions, y)]
+accuracy = sum(correct) / len(correct)
+print(f"Accuracy: {accuracy * 100}%")
+
```
Comment (glassnotes) [inline demonstrations/tutorial_qnn_module_torch.py:191]: Same as keras tutorial, would add some comments about the results to flow better into next section.

**Review by glassnotes — COMMENTED** [review]
These read nicely, and I think the usage examples make it very clear how to incorporate the quantum layers for both cases.

Given that there is a significant amount of overlap, is it possible to have one tutorial with a toggle at the top that will switch between keras/pytorch versions?

File: demonstrations/tutorial_qnn_module_tf.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
+"""
+Turning quantum nodes into Keras Layers
+=======================================
+
+.. meta::
+    :property="og:description": Learn how to create hybrid ML models in PennyLane using Keras
+    :property="og:image": https://pennylane.ai/qml/_images/Keras_logo.png
+
```
[the remark this answers is not in the document]
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_tf.py:8]: Thanks, that's a great idea! I've added.

File: demonstrations/tutorial_qnn_module_torch.py (hunk 17 above)
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_torch.py:8]: Thanks, added!

File: demonstrations/tutorial_qnn_module_tf.py [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+###############################################################################
+# Note that there are more advanced combinations of optimizer and loss function, but here we are
+# focusing on the basics.
+#
+# The model is now ready to be trained!
+
+X = X.astype("float32")
+y_hot = y_hot.astype("float32")
+fitting = model.fit(X, y_hot, epochs=6, batch_size=5, validation_split=0.25, verbose=2)
+
```
[the remark this answers is not in the document]
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_tf.py:161]: Yes good point. These are good questions, but unfortunately the accuracy isn't that great and I expect the network to be outperformed by a typical classical classifier.

This is because the hybrid model is so basic and we haven't really thought about hyperparameters, it's more of a proof of principle.

I've tried to add a few sentences though just to make the section ending less abrupt (af99ebc36564e3003f69a6e75cebd723f46f689f).

File: demonstrations/tutorial_qnn_module_torch.py (hunk 18 above)
Comment (trbromley) [reply demonstrations/tutorial_qnn_module_torch.py:195]: Thanks, yes I've added a few sentences (af99ebc36564e3003f69a6e75cebd723f46f689f).

**Review by trbromley — COMMENTED** [review]
Thanks @glassnotes!

**Comment by trbromley** [comment]
@glassnotes

> Given that there is a significant amount of overlap, is it possible to have one tutorial with a toggle at the top that will switch between keras/pytorch versions?

I had a go at using [sphinx-tabs](https://github.com/executablebooks/sphinx-tabs), but although the tabs would appear it was not possible to click between tabs. Just had another try now and no luck. We could also look at something like toggle boxes. These tutorials were originally as one and then were subsequently separated, which I think overall is the best option if sphinx-tabs won't work.

**Comment by josh146** [comment]
> These tutorials were originally as one and then were subsequently separated, which I think overall is the best option if sphinx-tabs won't work.

Yep, @glassnotes if you go through the commits you can see a version where it was a single tutorial. The problem with it being a single tutorial was it was a bit hard to follow; even the text paragraphs between the code blocks was subject to repetition and back and forth, and was hard to follow.

Further, most readers won't want to see both the torch _and_ the TF solutions; they will already have a framework of choice, and will be looking for one or the other.

**Comment by glassnotes** [comment]
> > These tutorials were originally as one and then were subsequently separated, which I think overall is the best option if sphinx-tabs won't work.
> 
> Yep, @glassnotes if you go through the commits you can see a version where it was a single tutorial. The problem with it being a single tutorial was it was a bit hard to follow; even the text paragraphs between the code blocks was subject to repetition and back and forth, and was hard to follow.
> 
> Further, most readers won't want to see both the torch _and_ the TF solutions; they will already have a framework of choice, and will be looking for one or the other.

Ahh fair enough. I was thinking like a global button or tab selection only once at the top of the page, rather than interleaving; but if sphinx-tabs isn't playing nice, it is better to just split in two then.

**Review by glassnotes — APPROVED** [review]

**Review by mariaschuld — APPROVED** [review]
